[쇼핑 장바구니 3, 4단계] 비비 미션 제출합니다

rosemin928 opened the pull request:
안녕하세요 코니!
3, 4단계를 구현하다가 리뷰 요청이 너무 많이 늦어져버렸네요...😢
3, 4단계 리뷰도 잘부탁드립니다!! 🙇‍♀️

# 셀프 체크리스트
- [x] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?
- 기능 구현에 집중하다가 테스트를 많이 작성하지 못했습니다. 특히 mockWebServer 테스트 부분은 거의 작성하지 못했습니다... 그리고 현재 작성되어 있는 부분도 이해하고 작성했다기보다 이게 맞나...? 생각하면서 chatGPT와 함께 작성했습니다. 이후 더 공부해보도록 하겠습니다...!!!

## 스크린샷
https://github.com/user-attachments/assets/8d52f7f5-5660-47a3-b1ab-467a597f2fde

rosemin928 committed:
refactor: 버튼 디자인 수정

rosemin928 committed:
feat: 상품 상세 페이지에서 장바구니 수량에 따른 담기 기능 구현

rosemin928 committed:
test: 불필요한 viewModel 생성 제거

rosemin928 committed:
test: 누락된 테스트 추가

rosemin928 committed:
docs: step3 상태 관리 기능 요구사항 작성

rosemin928 committed:
refactor: getProducts private 메서드로 변경

rosemin928 committed:
refactor: viewModel 접근제한자 수정

rosemin928 committed:
refactor: 토스트 메시지 이벤트 처리 SingleLiveData 이용하도록 수정

rosemin928 committed:
feat: + 버튼 누르면 장바구니에 수량 추가되는 기능 구현

rosemin928 committed:
refactor: 클릭리스너 바인딩해서 쓰도록 수정

rosemin928 committed:
refactor: 초기 수량 0으로 수정

rosemin928 committed:
feat: Cart 생성

rosemin928 committed:
feat: 장바구니에 상품 수량이 1개 이상 들어있을 때 장바구니 버튼 보이게 하는 기능 구현

rosemin928 committed:
feat: 장바구니 + 버튼 눌렀을 때 장바구니에 추가되는 기능 구현

rosemin928 committed:
feat: 장바구니 - 버튼 눌렀을 때 장바구니에서 줄어드는 기능 구현

rosemin928 committed:
feat: 상품의 수량 UI가 자동 업데이트 되는 기능 구현

rosemin928 committed:
feat: 상품 상세 페이지에서 상품 추가 이후 UI 갱신 기능 구현

rosemin928 committed:
refactor: 장바구니 담기 버튼을 눌렀을 때만 상품이 담기도록 수정

rosemin928 committed:
fix: 상품 추가를 할 때 가격 업데이트 되지 않도록 수정

rosemin928 committed:
feat: 장바구니에서 상품 추가/제거 기능 구현

rosemin928 committed:
refactor: resultCode enum 클래스로 관리

rosemin928 committed:
feat: Carts 일급컬렉션 구현

rosemin928 committed:
feat: 장바구니에 총 수량 아이콘 보이는 기능 구현

rosemin928 committed:
docs: 상태 관리 요구사항 완료 체크

rosemin928 committed:
docs: step4 기능 요구사항 작성

rosemin928 committed:
feat: HistoryRepository 생성

rosemin928 committed:
feat: 최근 본 상품 저장하는 기능 구현

rosemin928 committed:
fix: 더보기 버튼 깜빡이는 오류 수정

rosemin928 committed:
refactor: 더보기 버튼 span 수정

rosemin928 committed:
feat: 최근 본 상품 구현

rosemin928 committed:
refactor: 최근 본 상품 UI 갱신 기능 구현

rosemin928 committed:
feat: 마지막으로 본 상품 레이아웃 구현

rosemin928 committed:
feat: 마지막으로 본 상품 보여주는 기능 구현

rosemin928 committed:
refactor: 더 이상 볼 수 있는 상품이 없을 때 더보기 버튼이 뜨지 않게 수정

rosemin928 committed:
fix: 최근 본 상품 목록에 이미 해당 상품이 있을 때 마지막으로 본 상품 갱신 안 되는 문제 수정

rosemin928 committed:
refactor: 최근 본 상품 누르면 상세 페이지로 넘어가는 기능 구현

rosemin928 committed:
feat: 마지막으로 본 상품 누르면 해당 상품의 상세 페이지로 넘어가는 기능 구현

rosemin928 committed:
refactor: application 단위에서 repository 주입

rosemin928 committed:
feat: mockWebServer 구현

rosemin928 committed:
docs: step4 기능 구현 완료 체크

rosemin928 committed:
refactor: application 주입 수정

rosemin928 committed:
test: mockServer 테스트 구현

rosemin928 committed:
refactor: getAllItemsSize 안드로이드 의존성 제거

rosemin928 committed:
fix: 장바구니에 아이템 없을 때 navigate 안 되는 문제 수정

rosemin928 committed:
refactor: x 버튼 눌렀을 때 아이템 전체 삭제되도록 수정

rosemin928 committed:
fix: 장바구니에서 상품 수량 변경했을 때 UI 갱신 안 되는 문제 수정

rosemin928 committed:
fix: 장바구니에서 전체 삭제 이후 UI 갱신 안 되는 문제 수정

rosemin928 committed:
refactor: 장바구니에 0개일 때 아이콘 안 보이게 수정

rosemin928 committed:
test: 테스트 수정

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/goodsdetails/GoodsDetailsActivity.kt:
```diff
@@ -62,15 +60,68 @@ class GoodsDetailsActivity : AppCompatActivity() {
+                putExtra("HISTORY_ID", cart.id)
+            },
+        )
+    }
```
`setResult` 가 여기서 호출이 되는 것이 맞을까요?!..

rosemin928 replied:
상품 상세 페이지로 진입하자마자 저장하는게 맞는 것 같아서 `sendRecentHistory()`라는 함수를 만들고, 해당 로직의 실행 위치를 변경했습니다!

[refactor: 상품 상세 화면 진입 시 최근 본 상품 id 전달하는 로직 위치 변경](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/d9088a2827612619c83b105b0b3585742fb2e330)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/goodsdetails/GoodsDetailsActivity.kt:
```diff
@@ -62,15 +60,68 @@ class GoodsDetailsActivity : AppCompatActivity() {
+                ResultCode.GOODS_DETAIL_INSERT.code,
+                Intent().apply {
+                    putExtra("GOODS_ID", cart.id)
+                    putExtra("GOODS_QUANTITY", viewModel.cart.value?.quantity)
```
전반적으로 하드코딩된 String 들이 보이는데 다음 리뷰 요청때는 관리가 된다면 좋을 것 같네요

rosemin928 replied:
급하게 구현하다 보니 신경쓰지 못한 부분들이 많네요... 전부 상수화해주었습니다!

[refactor: intent 키 상수화](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/6dd0b40ecc3efac7bcf5cf0e7d2784ccbefef6ce)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/goodsdetails/GoodsDetailsViewModel.kt:
```diff
@@ -3,23 +3,86 @@ package woowacourse.shopping.feature.goodsdetails
+    private val _isLastViewedVisible = MutableLiveData<Boolean>()
+    val isLastViewedVisible: LiveData<Boolean> get() = _isLastViewedVisible
+    private val _isSuccess = MutableSingleLiveData<Unit>()
+    val isSuccess: SingleLiveData<Unit> get() = _isSuccess
```
무엇이 `success` 했다는 것이 명확한 변수명일까요? `fail` 도 그렇고 다시 고민해보시기 바랍니다.

rosemin928 replied:
각각의 상태를 하나의 LiveData로 관리하기 위해 sealed class 형태로 관리하도록 수정했습니다.
그 과정에서 더 명확한 변수명을 사용하기 위해 `cartInsertEvent`라는 이름으로 변수명을 변경했습니다!

[refactor: 액티비티 상태 관리 클래스 추가](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/c29ab4022e3c6ac2cdfc15d5ff9a447c77064c9e)
[refactor: 상품 상세 LiveData 변수명 변경](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/cd97d483d941c0a4538f8c25cb24d208737e338e)
[refactor: _cartResultEvent 네이밍 변경](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/a3da406ccdab802a0d0f12024cd3d29ee43ad32b)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -1,42 +1,105 @@
+
+            _items.postValue(updatedItems)
+        }
+        Log.d("123451", "dd")
```
정말로 많이 급했나보군요

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -45,7 +108,56 @@ class GoodsViewModel : ViewModel() {
         return dummyGoods.subList(fromIndex, toIndex)
     }
 
+    private fun loadItems() {
```
`loadItems` 에서 너무 많은 일을 하고 있는 것 같은데 한 번 좀 분리를 해봤으면 좋겠네요.

그리고 `_totalQuantity` 에서 `combinedItems` 만 계산하는게 아니라 전체 아이템 목록에서 계산을 안 해도 상관이 없을까요?

rosemin928 replied:
`_totalQuantity`에서 전체 아이템 목록의 총 수량을 계산해야 더보기 버튼을 눌렀을 때도 총 수량이 유지가 되겠네요...!

`loadItems`에서 histories와 carts를 불러오는 로직을 분리하고 두 목록을 합쳐주고 관리하는 일만 수행하도록 수정하고 `updateCombinedItems`로 함수명을 변경해주었습니다!

[refactor: 상품 목록 로딩 로직 개선](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/c4b50cee36982444ccb31932ecd4e04a6f3b88ec)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -1,42 +1,105 @@
+    private val historyRepository: HistoryRepository,
+) : ViewModel() {
+    private val _items = MutableLiveData<List<Any>>()
+    val items: LiveData<List<Any>> get() = _items
```
`Any`로 들어올 수 있군요... 아마 급해서 이렇게 된 것 같은데 

Goods 로 지금 들어올 수 있는 것은 

```
1. 최근 본 상품
2. 상품 목록
```

이것만으로도 충분히 sealed로 만들어서 관리가 가능할 것 같네요

rosemin928 replied:
```kotlin
sealed class GoodsItem {
    data class History(
        val histories: List<Cart>,
    ) : GoodsItem()

    data class Product(
        val cart: Cart,
    ) : GoodsItem()

    object LoadMore : GoodsItem()
}
```

최근 본 상품, 상품 목록, 더보기 버튼까지 sealed class로 만들어서 관리하도록 수정했습니다!

[refactor: GoodsItem sealed class 추가 및 GoodsAdapter ListAdapter로 변경](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/2bbbe95621ebd7d24749ebbb087fc062b38a4874)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -1,42 +1,105 @@
     fun addPage() {
         page++
-        loadGoods()
+        loadItems()
```
지금 보니 `addPage` 에서 `loadItems` 까지 한다는 것이 좀 어색하게 느껴지는데 어떻게 해보면 좋을까요?

> 그리고 지금 page를 먼저 ++ 하고 있는데 만약 `loadItems` 를 불러오는데 실패한 경우에도 page가 늘어나지 않을까요?

```
addPage() // page 1
addPage() // page 2 (loadItems 실패!)
addPage() // page 3 (3 을 불러오는게 맞을까)
```

rosemin928 replied:
`page`를 `loadCarts()`가 성공하면 업데이트하도록 수정했습니다!
addPage에서 `loadCarts()`를 하지 않을 방법을 계속 생각해봤으나 잘 떠오르지 않아 우선 carts와 history를 불러오는 로직을 분리한 뒤, cart만 불러오도록 수정해두었습니다.

[refactor: 상품 목록 로딩 및 페이징 로직 개선](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/91b02da79a67f372a0a2701ec03201856567fd2b)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -45,7 +108,56 @@ class GoodsViewModel : ViewModel() {
+        val currentItems = _items.value.orEmpty()
+
+        historyRepository.getAll { histories ->
+            cartRepository.getAll { cartsResult ->
```
지금은 `history`를 다 불러오고 `cart` 을 불러오고 있는데요.

이러면 `cart` 를 불러오는 것을 비동기로 처리하는 것이 살짝 아쉽기도 하고 `cart` 를 불러올 때 `history` 는 관계가 없으니 따로따로 갖고 올 수 있지 않을까요?

그리고 분리를 한다면 2가지 선택을 할 수 있을 것 같습니다.

```
1. 각각의 LiveData를 만든다.
2. 하나의 LiveData로만 update를 한다.
```

1번의 장점의 경우 각 아이템에 대한 삭제 혹은 관리가 용이하다는 장점이 있는데요 단 `Adapter`에 사용하기에는 어울리지 않기에 `LiveData` `N`개를 합치는 `MediatorLiveData` 를 학습하고 적용을 해본다면 더 좋을 것 같네요.

rosemin928 replied:
`histories`와 `carts`라는 각각의 LiveData를 만들고 items를 `MediatorLiveData`로 만들어 합쳐주었습니다!

[refactor: 상품 목록 로딩 로직 개선](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/c4b50cee36982444ccb31932ecd4e04a6f3b88ec)

rosemin928 replied:
**MediatorLiveData** : 일반 MutableLiveData처럼 동작하지만, 다른 LiveData를 addSource로 연결해 관찰하면서 가공된 결과를 가질 수 있다. 보통 여러 LiveData 값을 하나로 묶고 싶을 때 사용한다.

```kotlin
private val _items = MediatorLiveData<List<GoodsItem>>()
val items: LiveData<List<GoodsItem>> get() = _items

...

_items.addSource(histories) { updateCombinedItems() }
```

MediatorLiveData _items에 `histories`라는 LiveData를 소스로 등록하고, `histories`의 값이 변경될 때마다 `updateCombinedItems()`가 호출된다.

lee-ji-hoon commented on app/src/main/res/layout/activity_goods_details.xml:
```diff
@@ -51,45 +48,55 @@
-            android:id="@+id/tv_goods_details_price"
-            android:layout_width="wrap_content"
-            android:layout_height="wrap_content"
+        <woowacourse.shopping.feature.CustomCartQuantity
+            android:id="@+id/custom_cart_quantity"
```
잘못됐다는 의미는 아니며, `include` 로 `layout` 을 넣는 방식도 있었을 텐데 `custom layout` 을 선택한 이유는 어떤 것일까요?

rosemin928 replied:
커스텀 뷰 클래스를 만드는 쪽이 `setCount()`나 `setClickListener()` 등과 같은 메서드를 관리할 때 더 용이하다고 생각하여 해당 방식을 사용했습니다!
만약 뷰만 재사용하는 것이라면 `include`를 사용했을 것 같은데, 클릭 이벤트(상품 더하기, 상품 빼기 등)나 수량 바인딩 등이 필요하기 때문에 이를 더 보기 쉽게 관리하기 위해 커스텀 뷰 클래스를 사용했습니다...!

lee-ji-hoon commented on README.md:
```diff
@@ -11,11 +12,27 @@
+- [x] 최근 본 상품이 있는 경우 상품 목록 상단에서 10개까지 확인할 수 있다.
+- [x] 마지막으로 본 상품 1개를 상품 상세 페이지에서 확인할 수 있다.
+  - [x] 마지막으로 본 상품을 선택했을 때는 마지막으로 본 상품이 보이지 않는다.
+  - [ ] 마지막으로 본 상품 페이지에서 뒤로 가기를 하면 상품 목록으로 이동한다.
```
구현은 된 것 같은데 체크는 안 되어 있네요 👀

rosemin928 replied:
앗 바쁘게 하다가 체크를 놓쳤네요...🥲 수정했습니다!

lee-ji-hoon commented on app/src/androidTest/java/woowacourse/shopping/GoodsDetailsActivityTest.kt:
```diff
@@ -10,15 +10,16 @@ import androidx.test.platform.app.InstrumentationRegistry
 
     private val testGoods =
-        GoodsUiModel(
+        CartUiModel(
+            id = 0,
```
여기가 왜 `GoodsUiModel`에서 `CartUiModel`로 변경이 되었을까? 하고 살펴보니 `GoodsDetailsActivity`가 `newIntent` 에서 `CartUiModel` 을 받고 있네요. 상품 상세 화면으로 이동할 때 Goods가 아닌 Cart 아이템이 필요한 것은 어색하다는 생각이 드는데 비비는 어떻게 생각하시나요?

rosemin928 replied:
구현이 급해서 상품 상세 화면에 필요한 모든 것을 넘겨주려고 cart를 그대로 사용하게 되었었는데 확실히 어색하다는 생각이 드네요 🥲
이 부분을 `goodsId`를 넘겨주고 그 `goodsId`로 상품을 찾아서 만약 카트에 상품이 있다면 수량까지 표시해주고, 없다면 goods의 기본 정보를 가져와서 사용하도록 수정해주었습니다!

[refactor: History 도메인 모델 생성에 따른 로직 수정](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/0173a6b523356aec447b1e3e97e8cb7c77876b30)
[refactor: GoodsDetailsActivity에서 cartId를 goodsId로 변경](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/54b8f9e4d253af38b925b38008bd5734dd855977)

rosemin928 replied:
수정하는 부분이 커져서 커밋에서 변경사항을 보기가 쉽지가 않네요 죄송합니다...

[refactor: History 도메인 모델 생성에 따른 로직 수정](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/0173a6b523356aec447b1e3e97e8cb7c77876b30) 

위 커밋의 `GoodsDetailsViewModel`과 `GoodsDetailsActivity`의 변경사항입니다!

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/application/ShoppingApplication.kt:
```diff
@@ -0,0 +1,40 @@
+class ShoppingApplication : Application() {
+    private val database by lazy { ShoppingDatabase.getDatabase(applicationContext) }
+
+    val goodsFactory by lazy {
+        ViewModelFactory {
+            GoodsViewModel(
+                CartRepositoryImpl(database.cartDao()),
+                HistoryRepositoryImpl(database.historyDao()),
+            )
+        }
+    }
```
주입은 Repository / DataBase 정도만 하고 `viewModel` 은 Activity에서 `ViewModelFactory` 같은 것으로 생성이 가능했을 텐데 Application에서 하기로 결정한 이유가 있을까요?

rosemin928 replied:
Activity마다 factory를 만들어주는 코드가 반복되고, 보기에도 좀 더 복잡할 것 같다 생각하여 Application에서 viewModel까지 전부 생성해주었었습니다.
하지만 현재는 viewModel마다 factory를 만들어주는 코드가 달라서 이점이 없기도 하고, 같은 viewModel을 여러 화면에서 사용할 수 없겠다는 단점도 있을 것 같네요...

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/data/Mapper.kt:
```diff
@@ -0,0 +1,17 @@
+import woowacourse.shopping.domain.model.Goods
+
+fun Cart.toEntity(): CartEntity =
+    CartEntity(id = goods.id, name = goods.name, price = goods.price, thumbnailUrl = goods.thumbnailUrl, quantity = quantity)
```
`Mapper` 를 전부 하나로 모은다면 `Constant` 처럼 상수들을 모아두는 것과 별반 다를 것이 없다고 생각을 하는데 원래는 나눠져 있던 것으로 기억을 하는데 어떤 이유로 합치게 된 것일까요?

rosemin928 replied:
Mapper 함수가 그렇게 많지 않아 모아 두는건 어떨까 생각했었는데 분리하는 편이 관리하기 더 편할 것 같네요! 수정했습니다!!
더불어 원래 `feature.util`에 있던 `Mapper`가 있었는데, 이름이 적절하지 않다 생각하여

```kotlin
fun updateQuantity(newQuantity: Int): Cart = this.copy(quantity = newQuantity)
```

위 함수는 Cart의 확장함수이기 때문에 Cart 내부로 옮겨주고

```kotlin
fun List<GoodsItem.Product>.updateCartQuantity(
    id: Long,
    newQuantity: Int,
): List<GoodsItem.Product> =
    this.map { item ->
        if (item.cart.goods.id == id) {
            GoodsItem.Product(item.cart.copy(quantity = newQuantity))
        } else {
            item
        }
    }
```

위 함수는 `GoodsExtensions`라는 파일을 만들어 관리하도록하여 제거했습니다!

[refactor: CartMapper 파일 분리](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/61fe33e1acfeea05b0bdda9ef02bd38030e1d4c2)
[refactor: Mapper.kt 파일 삭제](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/abf1a97925254e888ca303d6522a1e2c6bd28378)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/data/cart/CartDao.kt:
```diff
@@ -0,0 +1,55 @@
+    fun insertAndUpdateQuantity(cartEntity: CartEntity) {
+        insertAll(cartEntity)
+        update(cartEntity.copy(quantity = 1))
+    }
```
DAO에서 이 동작이 꼭 필요할까? 라는 생각이 드는데요.

장바구니에 처음 추가된 아이템의 수량이 1이다. 이것은 Entity에 기본값으로 존재를 하거나 다른 곳에서 이미 세팅이 되어서 DB에서는 저장만 하면 되지 않았을까 하는데 비비는 어떻게 생각하실까요?

rosemin928 replied:
Entity에 기본값으로 존재를 하는게 맞는 것 같네요...! 구현 당시에 생각을 못했던 것 같습니다. 수정했습니다!

[refactor: CartEntity 기본 수량 1로 설정 및 insert 로직 수정](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/b56a3d3e568a7606a1771450257e199bc3b2e3bd)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/data/cart/server/CartService.kt:
```diff
@@ -0,0 +1,15 @@
+import woowacourse.shopping.domain.model.Cart
+import woowacourse.shopping.domain.model.Carts
+
+interface CartService {
```
Cart는 이미 Room으로 구현이 되어 있는데 왜 Service가 있지? 하고 찾아보니 `TestCartServiceImpl` 만 존재하고 테코에서도 선언만 되어 있고 사용은 안 하는 것 같은데 어떤 의도일까요?..

rosemin928 replied:
프로그래밍 요구사항으로
```
MockWebServer를 사용하여 테스트 환경을 만든다.
```
라는 항목이 있어 테스트 환경에서만 사용을 하도록 구현하려 했습니다.
하지만 mockWebServer를 사용하는 것이 익숙하지 않아 제대로 사용하지 못했던 것 같습니다...😢
너무 리뷰 반영 기간이 길어지는 것 같아 해당 부분 리뷰 요청 이후에 다시 제대로 공부하여 수정하도록 하겠습니다!!

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/data/history/repository/HistoryRepositoryImpl.kt:
```diff
@@ -0,0 +1,35 @@
+    override fun insert(cart: Cart) {
+        thread {
+            val isNew = dao.findById(cart.goods.id) == null
+            if (isNew && dao.getAll().size >= 10) {
```
10개 이상이면 `oldest` 를 지운다. 이것은 비즈니스적인 정보이지 않았을까요? 그러면 이것이 비즈니스 정보면 어디서 관리를 해야 할까? 부터 고민해보면 생각보다 어려울 수 있는데 비비가 한 번 고민을 해보면 좋을 것 같네요.

rosemin928 replied:
10개 이상이면 `oldset`을 지운다는 로직은 비즈니스 정보가 맞는 것 같습니다.
이 정보를 어디서 관리해야할지 알아보는 과정에서 `usecase`에 대해 알게 되었는데, 이에 대해 말씀하시려던 것이 맞을까요...!?

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/data/history/repository/HistoryRepositoryImpl.kt:
```diff
@@ -0,0 +1,35 @@
+            val cart = dao.getAll().map { it.toDomain() }
+            callback(cart)
+        }
+    }
```
전반적으로 `callback`이라는 네이밍을 많이 사용 중인데. 데이터를 불러오는 데 성공했을 때 실행할 lambda라는 이름을 좀 더 명확하게 해 보면 어떨까요?

> 더불어서 지금은 Room이라서 희박한 확률이긴 하지만 error가 났을 때도 고민해보시기 바랍니다.

rosemin928 replied:
`callback`이라는 이름을 `onSuccess`라는 이름으로 성공했을 때 실행하는 lambda인 것을 좀 더 명확하게 보여주도록 수정했습니다!
만약 에러가 났을 때에는 `onFailure` 혹은 `onError`라는 lambda를 통해 에러 처리를 해주어야 할 것 같습니다...!

[refactor: 콜백 함수명 onSuccess로 변경](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/96a9f5dc89d6c401628d69d7de592b5cfda395c9)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/data/history/repository/HistoryRepository.kt:
```diff
@@ -0,0 +1,11 @@
+
+import woowacourse.shopping.domain.model.Cart
+
+interface HistoryRepository {
```
지금 보면 `domain model : repository : entity = 1 : 1 : 1` 이렇게 관리가 되고 있늗네요.

[공식문서](https://developer.android.com/topic/architecture/data-layer#architecture)의 data layer repository에 대해서는 아래와 같이 네이밍을 권장하고 있는데요.

```
You should create a repository class for each different type of data you handle in your app
```

그런 측면에서 바라봤을 때 `History` 가 올바른 네이밍일지 고민해보시면 좋을 것 같네요

rosemin928 replied:
구현을 할 때 거의 모든 화면에서 Cart를 넘기도록 했었는데, 최근 본 상품에서는 이름과 사진만 보이기 때문에 History라는 도메인 모델을 구현하여 사용하는게 맞다고 느껴져 다음과 같은 모델을 만들었습니다!

```kotlin
data class History(
    val id: Long,
    val name: String,
    val thumbnailUrl: String,
)
```

HistoryRepository에서도 이제 Cart가 아닌 History를 받아 사용하고, 해당 정보를 넘겨받는 화면에서는 History의 아이디로 상품을 찾아서 보여주도록 변경했습니다!

[feat: 최근 본 상품 도메인 모델 추가](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/42275711616b23c67a17069607cb7f30dac49722)
[refactor: History 도메인 모델 생성에 따른 로직 수정](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/0173a6b523356aec447b1e3e97e8cb7c77876b30)
[feat: ID를 통해 장바구니 조회 기능 추가](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/70a88a7fa60b017906ef63b85f8dbf3968173a5a)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/domain/model/Cart.kt:
```diff
@@ -0,0 +1,8 @@
+
+data class Cart(
+    val goods: Goods,
+    var quantity: Int,
+) {
+    val totalPrice: Int get() = goods.price * quantity
```
`totalPrice` 는 단순히 생각하면 `uiModel`에서 `goods.price` 랑 `quantity`를 곱하면 해결이 되는 사항이라서 거기서 할 수 있었을텐데 잘 관리해주셨습니다.

그런데 왜 `get()` 인가 하고 봤더니 `quantity` 가 var 이네요 👀

rosemin928 replied:
앗,, 처음에 다른 방식으로 구현을 했다가 이후 수정을 못한 것 같네요 😢
수정했습니다!!

[refactor: Cart의 quantity를 val로 변경](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/61eb6e13f3566654a00540785b6e9dca2f95f6b6)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/BindingAdapter.kt:
```diff
@@ -14,16 +14,31 @@ fun ImageView.loadImageFromUrl(url: String) {
 @BindingAdapter("visible")
 fun View.setVisible(visible: Boolean) {
-    this.visibility = if (visible) View.VISIBLE else View.GONE
+    visibility = if (visible) View.VISIBLE else View.GONE
```
제가 리뷰를 남겼는지 기억이 애매한데

```suggestion
    isVisible = visible
```

에서 `isVisible` 내부를 보셔도 좋고 저기 안에 있는 `View` 의 확장함수는 무엇이 있는지 한 번 보셔도 좋을 것 같네요

rosemin928 replied:
```kotlin
public inline var View.isVisible: Boolean
    get() = visibility == View.VISIBLE
    set(value) {
        visibility = if (value) View.VISIBLE else View.GONE
    }
```

이미 View의 확장함수가 있었군요...! 들어가서 보니까 다른 확장함수들도 많네요!!
나중에 미션을 하면서 써보겠습니다 감사합니다!!

[refactor: BindingAdapter에 isVisible 확장 함수 적용](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/af68917ba1b0a0686b1b865058537b09a4311f3d)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/cart/CartActivity.kt:
```diff
@@ -1,33 +1,33 @@
         binding.lifecycleOwner = this
-        binding.rvGoods.adapter = adapter
         binding.viewModel = viewModel
+        binding.rvGoods.adapter = adapter
```
이 내용도 setupAdapter에 있는 것도 좋아보이네요

rosemin928 replied:
수정했습니다!

[refactor: 장바구니 리사이클러뷰 어댑터 설정 위치 변경](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/a3eacdd777ea762fb1518fd6bb1cc9daba2f533c)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/cart/adapter/CartViewHolder.kt:
```diff
@@ -14,8 +15,19 @@ class CartViewHolder(
+                    cartClickListener.removeFromCart(cart)
+                }
+            },
+        )
```
init에서 ClickListener를 거는 방법과 bind에서 거는 방법 이렇게 주로 2가지가 존재하는데 둘의 차이점이 어떨 것 같나요?

`ViewHolder` 의 재활용 측면에서 고민해보면 답이 나옵니다.

rosemin928 replied:
`ViewHolder`가 재사용되고 있기 때문에 클릭 리스너도 init으로 한 번만 걸어주면 계속 사용될 수 있기 때문에 재활용 측면에서 더 좋을 것 같네요! 수정했습니다!

[refactor: 카트 상품 클릭 이벤트 로직 수정](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/508adda3ed5abe1116f21956cd5a0276211c2548)

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsActivity.kt:
```diff
@@ -4,70 +4,148 @@ import android.content.Intent
+    private val viewModel: GoodsViewModel by viewModels {
+        (application as ShoppingApplication).goodsFactory
+    }
+    private lateinit var activityResultLauncher: ActivityResultLauncher<Intent>
```
`ActivityResultLauncher`를 잘 활용해주셨습니다 👍

lee-ji-hoon commented on app/src/main/java/woowacourse/shopping/feature/goods/adapter/GoodsAdapter.kt:
```diff
@@ -4,38 +4,109 @@ import android.view.LayoutInflater
-        val newItemCount = newItems.size
+    fun setItems(newItems: List<Any>) {
+        val oldHistory = items.firstOrNull()?.takeIf { it is List<*> && it.all { h -> h is Cart } } as? List<*>
+        val newHistory = newItems.firstOrNull()?.takeIf { it is List<*> && it.all { h -> h is Cart } } as? List<*>
```
![Screenshot_20250526_221423](https://github.com/user-attachments/assets/a5e6243c-7742-414b-9d04-0d2683d7ce72)

위 처럼 최근 상품이 없을 때도 보이고, `NestedScrollView`도 안 쓰고 구현 했길래 `ConcatAdapter` 인가? 하고 보니까 그것은 아니네요. 시간이 없어서 구현을 못한 것 같기도 한데, 개선이 된다면 좋을 것 같네요.

rosemin928 replied:
해당 부분 리스트 어댑터로 변경하며 수정했습니다!

[refactor: GoodsItem sealed class 추가 및 GoodsAdapter ListAdapter로 변경](https://github.com/woowacourse/android-shopping-cart/pull/127/commits/2bbbe95621ebd7d24749ebbb087fc062b38a4874)

lee-ji-hoon reviewed (changes requested):
비비 안녕하세요! 우선 3,4단계 고생 많았습니다.

`NestedScrollView` 를 안 쓰기 위해서 여러 노력을 해주셨고, `ActivityResult` 를 사용해서 관리를 하는 부분에서 가이드를 충실히 잘 따라가고 있다고 보이네요. 단 전반적으로 시간이 없어서 덜 처리됐거나 구현 부분에서 아쉬운 부분들이 있는데요. 이미 비비가 충분히 알 것이라고 생각은 하지만 전반적으로 리뷰를 남겨뒀습니다.

한 번 천천히 고민을 해보고 리뷰를 반영해 보거나 의견을 남겨보시면 좋을 것 같네요.

> 코드래빗 제가 resolve 처리 안 한 것들은 한 번 읽어보시기 바랍니다.

---

혹시 너무 고봉밥이다 싶으면 어떤 리뷰에는 수도 코드 정도로만 남겨주셔도 됩니다.

rosemin928 committed:
build: 필요 없는 MockWebServer 의존성 제거

rosemin928 committed:
refactor: 상품 상세 화면 진입 시 최근 본 상품 id 전달하는 로직 위치 변경

rosemin928 committed:
refactor: intent 키 상수화

rosemin928 committed:
refactor: 액티비티 상태 관리 클래스 추가

- GoodsActivity와 GoodsDetailsActivity에서 State를 관찰하여 토스트 메시지 표시 로직 변경

rosemin928 committed:
refactor: 상품 목록 로딩 로직 개선

- `_items` LiveData를 `MediatorLiveData`로 변경하여 여러 데이터 소스(최근 본 상품, 장바구니 상품)를 결합

rosemin928 committed:
docs: 4단계 기능 구현 완료 체크

rosemin928 committed:
refactor: 상품 상세 LiveData 변수명 변경

rosemin928 committed:
refactor: GoodsItem sealed class 추가 및 GoodsAdapter ListAdapter로 변경

rosemin928 committed:
refactor: CartEntity 기본 수량 1로 설정 및 insert 로직 수정

rosemin928 committed:
refactor: 장바구니 리사이클러뷰 어댑터 설정 위치 변경

- 어댑터 설정 로직을 `setupAdapter()` 함수 내부로 이동

rosemin928 committed:
refactor: `_cartResultEvent` 네이밍 변경

rosemin928 committed:
refactor: Cart의 quantity를 val로 변경

rosemin928 committed:
refactor: BindingAdapter에 isVisible 확장 함수 적용

rosemin928 committed:
feat: FakeGoodsRepository 구현

rosemin928 committed:
feat: 최근 본 상품 도메인 모델 추가

rosemin928 committed:
refactor: History 도메인 모델 생성에 따른 로직 수정

rosemin928 committed:
feat: ID를 통해 장바구니 조회 기능 추가

rosemin928 committed:
refactor: Glide url을 nullable로 변경

rosemin928 committed:
refactor: GoodsDetailsActivity에서 cartId를 goodsId로 변경

- `cartId` 변수명을 `goodsId`로 변경하여 의미를 명확하게 함

rosemin928 committed:
refactor: CartMapper 파일 분리

- Cart 관련 매핑 함수를 CartMapper.kt 파일로 분리
- 기존 Mapper.kt 파일명을 HistoryMapper.kt로 변경

rosemin928 committed:
refactor: 사용하지 않는 UI 모델 제거

- CartUiModel 및 GoodsUiModel 제거

rosemin928 committed:
refactor: Mapper.kt 파일 삭제

- `Cart` 모델에 `updateQuantity` 함수 이동
- `GoodsItemExtensions.kt` 파일을 생성하여 확장 함수 이동

rosemin928 committed:
refactor: GoodsViewModel 상품 목록 업데이트 로직 개선

- `updateCombinedItems` 함수를 `updateHistories`와 `updateCarts`로 분리하여 각 데이터 소스 변경에 따른 UI 업데이트 로직을 명확히 함

rosemin928 committed:
refactor: 카트 상품 클릭 이벤트 로직 수정

rosemin928 committed:
refactor: 상품 목록 로딩 및 페이징 로직 개선

- `addPage()`: 다음 페이지 상품이 있을 경우에만 페이지를 증가시키고 UI를 업데이트하도록 변경

rosemin928 committed:
refactor: 콜백 함수명 onSuccess로 변경

rosemin928 commented:
안녕하세요 코니!
최대한 리뷰를 전부 반영해보려 했으나 테스트, 페이징 부분에서 제대로 반영하지 못한 부분도 있습니다... 죄송합니다 😢
하지만 리뷰 반영 기간이 너무 길어지는 것 같아 우선 리뷰 요청을 보냅니다...!

긴 시간동안 기다려주셔서 감사합니다!! 🙇‍♀️

lee-ji-hoon reviewed (approved):
비비 안녕하세요. 아쉽게 전부 마무리는 안 됐지만, 적용해 주신 부분들은 확인해 봤을 때 큰 문제가 없어서 이제 마무리를 하고자 합니다.

새로운 기술을 적용하면서 어려웠던 점이 많았던 것 같은데 지금의 경험들이 앞으로 쌓여서 새로운 것을 배울 때 어떻게 공부하고 적용해 나갈지 학습하는 시간이라고 생각하고 레벨 2를 마무리하면 좋을 것 같네요.

고생 많았습니다 비비
